sktime/annotation/stray.py

```diff
@@ -0,0 +1,283 @@
+
+        Parameters
+        ----------
+        X : np.ArrayLike
+            Data for anomaly detection (time series).
+        n : int
+            The number of rows remaining in X when NA's are removed.
+
+        Returns
+        -------
+        dict of index of outliers and the outlier scores
+        """
+        if len(X.shape) == 1:
+            nbrs = NearestNeighbors(n_neighbors=self.k + 1).fit(X.reshape(-1, 1))
+            distances, _ = nbrs.kneighbors(X.reshape(-1, 1))
+        else:
+            nbrs = NearestNeighbors(
+                n_neighbors=n if self.k >= n else self.k + 1,
+                algorithm=self.knn_algorithm,
+            ).fit(X)
+            distances, _ = nbrs.kneighbors(X)
+
+        if self.k == 1:
+            d = distances[:, 1]
+        else:
+            diff = np.apply_along_axis(np.diff, 1, distances)
+            d = distances[range(n), np.apply_along_axis(np.argmax, 1, diff) + 1]
+
+        out_index = self._find_threshold(d, n)
+        return {"idx_outliers": out_index, "out_scores": d}
+
+    def _find_HDoutliers(self, X):
+        """Detect Anomalies in High Dimensional Data.
+
+        Parameters
+        ----------
+        X : np.ArrayLike
+            Data for anomaly detection (time series).
+
+        Returns
+        -------
+        dict of anomalies and their corresponding scores
+        """
+        r = np.shape(X)[0]
+        idx_dropna = np.array([i for i in range(r) if not np.isnan(X[i]).any()])  # tag
+        X_dropna = X[
+            idx_dropna,
+        ]  # FIXME: might move this somewhere else?
+
+        X_dropna = np.apply_along_axis(self.normalize, 0, X_dropna)
+
+        n = np.shape(X_dropna)[0]
+        outliers = self._use_KNN(X_dropna, n)
+
+        # adjusted back to length r, for missing data
+        slice_ = [True if i in outliers["idx_outliers"] else False for i in range(n)]
+        idx_outliers = idx_dropna[slice_]  # index values from 1:r
+        outlier_bool = np.array([1 if i in idx_outliers else 0 for i in range(r)])
+
+        list_scores = outliers["out_scores"].tolist()
+        outlier_scores = np.array(
+            [list_scores.pop(0) if i in idx_dropna else np.nan for i in range(r)]
+        )
+
+        return {
+            "outlier_scores": outlier_scores,
+            "outlier_bool": outlier_bool,
+        }
+
+    def _fit(self, X, y=None):
+        """Find outliers using STRAY (Search TRace AnomalY).
+
+        Parameters
+        ----------
+        X : pd.DataFrame
+            Training data to fit model to (time series).
+        y : pd.Series, optional
+            Not used for this unsupervsed method.
+
+        Returns
+        -------
+        self :
+            Reference to self.
+        """
+        # remember X for transform
+        self._X = X
+
+        info_dict = self._find_HDoutliers(X)
+        self.score_ = info_dict["outlier_scores"]
+        self.y_ = info_dict["outlier_bool"]
```

Comment:
can you please look into `HampelFilter`, there we replace outliers with NaN or `bool`:
```
    return_bool : bool, optional
        If True, outliers are filled with True and non-outliers with False.
        Else, outliers are filled with np.nan.
```
Lets have same interface for both? Having NaN is convenient so user can directly apply an `Imputer` after to correct outliers. Whats do you think?

Reply:
I think for now its best to follow `HampelFilter` interface until we have agreed on annotation output, e.g. storing stuff in index. Then we can change both to have same interface?

Reply:
Can do - but I have a question: what if the input data has np.nan? How do we distinguish this from an anomaly when `return_bool = False`?

Reply:
I have changed to boolean flags for output, for now. 

Exploring the other option, introducing np.nan for the anomalous values, would mean input
```
    X = np.array(
        [
            [-1.20706575, -0.57473996],
            [0.27742924, -0.54663186],
            [1.08444118, np.nan],
            [-2.3456977, -0.89003783],
            [0.42912469, -0.4771927],
            [0.50605589, -0.99838644],
        ]
    )
```
gives expected boolean output 
`y = [False, False, False, True, True, True]`
the algorithm specifies any na values are assumed normal data.

Having an output that is the original data with anomalies -> na gives
``` 
X = np.array(
        [
            [-1.20706575, -0.57473996],
            [0.27742924, -0.54663186],
            [1.08444118, np.nan],
            [np.nan, np.nan],
            [np.nan, np.nan],
            [np.nan, np.nan],
        ]
    )
```
which changes dimensionality, and is a bit confusing because the second most important information are the corresponding anomaly scores
scores_ = `[0.5233413, 0.5233413, np.nan, 0.7248368, 0.6035040, 0.8704687]`

Do we still think the np.nan anomaly overwrite makes sense for this use case? @aiwalter @lmmentel

Reply:
I don't see any value in overriding the input data. How would that be used, and under what circumstances? For me in the context of time series we are flagging anomalous **time point** or **periods of time** therefore a boolean 1D array should be the standard output. If the algorithm can **in addition** flag individual points I would say that a boolean mask with the same dimensionality as the input would be a good choice since you can then filter the input data whichever way you want and get either only anomalous data or only normal data. If the algorithm provides even more information like anomaly scores then I would apply same logic.